java/src/test/java/org/wildfly/openssl/BasicOpenSSLCustomEngineTest.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.openssl;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * @author Heyuan Liu
+ */
+public class BasicOpenSSLCustomEngineTest {
+
+    @Test
```

Comment:
Something that I noticed was that the outcome of the tests depends on the order in which the test cases are run.

Because the second test below is expected to successfully use the system property and call `SSL#getInstance()`, this means that if the second test happens to be run before the first test, then `SSL#init` won't be run again and the first test will fail on line 33.

To fix this, you could specify that the first test must be run before the second test. This could be done by adding `@InSequence(1)` and `@InSequence(2)` annotations to the test methods (see `org.jboss.arquillian.junit.InSequence` for more details).

Reply:
Thanks for your comments. I will fix this.

Reply:
Arquillian isn't in project's maven dependency. Should I add this dependency or upgrade the Junit version from Junit 4 to Junit 5 (maybe there is API change to upgrade Junit5)? 

Or I separate those test methods to two test classes ?
Any comments about this?

Reply:
Good question, we haven't checked yet if it's easy to upgrade to JUnit 5. For now, it's fine to add the following dependency:

```
<dependency>
    <groupId>org.jboss.arquillian.junit</groupId>
    <artifactId>arquillian-junit-core</artifactId>
    <version>1.6.0.Final</version>
    <scope>test</scope>
</dependency>
```

Reply:
I still see random order issues if I launch the new test alone in my env:

```
mvn clean install -Dorg.wildfly.openssl.path=/home/rmartinc/apps/openssl-1.1.1f/lib  -Dorg.wildfly.openssl.engine=qatengine -Dtest=org.wildfly.openssl.BasicOpenSSLCustomEngineTest
...
[INFO] Running org.wildfly.openssl.BasicOpenSSLCustomEngineTest
[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.093 s <<< FAILURE! - in org.wildfly.openssl.BasicOpenSSLCustomEngineTest
[ERROR] testUnknownEngine(org.wildfly.openssl.BasicOpenSSLCustomEngineTest)  Time elapsed: 0.003 s  <<< FAILURE!
java.lang.AssertionError: Expected ExceptionInInitializerError not thrown
```

My feeling is that only junit4 is being used and the arquillian `InSequence` annotation is doing nothing. In junit4 (but since 4.11) we can use `@FixMethodOrder(MethodSorters.NAME_ASCENDING)` and call the methods `test1UnknownEngine` and `test2RDRANDEngine` (or similar) to have the order we want. But that needs an upgrade of junit to last `4.13.2` (which is not bad because this project is using a very old junit 4.8.2). I just see a problem in `TestAllMethodsImplemented.java` with the upgrade, but is easily fixable using `MatcherAssert.assertThat`.

@heyuanliu-intel @fjuma WDYT about using `FixMethodOrder` and upgrading junit4?

The rest  of the PR LGTM. When I reviewed the natives PR I used this one too and everything was OK for me.

Reply:
Since @rmartinc has the changes for upgrading to JUnit 4.13.2 done locally, he's going to submit a PR for that and we'll get that merged.

@heyuanliu-intel You'll then be able to rebase and use the `@FixMethodOrder` annotations as @rmartinc mentioned.

Thank you both!

Reply:
@heyuanliu-intel We just merged @rmartinc's PR that updates to JUnit 4.13.2 (https://github.com/wildfly-security/wildfly-openssl/pull/130) so you should be able to rebase now.